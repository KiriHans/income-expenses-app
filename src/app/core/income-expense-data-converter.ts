import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { IncomeExpense } from './models/income-expenses.model';

export const incomeExpenseConverter: FirestoreDataConverter<IncomeExpense> = {
  toFirestore: (incomeExpense: IncomeExpense) => {
    const { description, amount, type, uid } = incomeExpense;
    return { description, amount, type, uid };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<IncomeExpense>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new IncomeExpense(data.description, data.amount, data.type, data.uid);
  },
};
