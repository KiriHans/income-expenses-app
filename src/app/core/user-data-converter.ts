import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { UserApp } from './models/users.model';

export const userConverter: FirestoreDataConverter<UserApp> = {
  toFirestore: (user: UserApp) => {
    const { displayName, email } = user;
    return { displayName, email };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<UserApp>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new UserApp(data.id, data.displayName, data.email);
  },
};
