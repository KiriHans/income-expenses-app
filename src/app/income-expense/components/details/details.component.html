<div class="row">
  <div class="col-12">
    <div class="card">
      <div class="card-body">
        <h4 class="card-title">Income expenses details</h4>

        <table class="table">
          <thead>
            <tr>
              <th>Description</th>
              <th>Amount</th>
              <th>Type</th>
              <th>Actions</th>
            </tr>
          </thead>

          <tbody>
            <tr *ngFor="let item of incomeExpenseItems$() | sortIncomeExpense">
              <td>{{ item.description }}</td>
              <td>{{ item.amount }}</td>
              <td
                [ngClass]="{
                  'text-success': item.type === 'income',
                  'text-danger': item.type === 'expense'
                }"
              >
                {{ item.type }}
              </td>
              <td>
                <button (click)="delete(item.id)" type="button" class="btn btn-danger">
                  Delete
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
