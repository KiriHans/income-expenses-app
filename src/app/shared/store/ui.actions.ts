import { createActionGroup, emptyProps } from '@ngrx/store';

export const UiActions = createActionGroup({
  source: 'UI',
  events: {
    'Load UI': emptyProps(),
    'Stop Load UI': emptyProps(),
  },
});
