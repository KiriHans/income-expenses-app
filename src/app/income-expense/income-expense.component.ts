import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'app-income-expense',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './income-expense.component.html',
  styles: [],
})
export class IncomeExpenseComponent {}
