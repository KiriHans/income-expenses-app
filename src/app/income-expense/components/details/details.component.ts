import { Component, OnInit, Signal } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Store } from '@ngrx/store';
import { IncomeExpense } from 'src/app/core/models/income-expenses.model';
import { incomeExpenseFeature } from 'src/app/store/reducers/income-expense.reducer';
import { DetailsActions } from 'src/app/store/actions/income-expense.actions';
import { SortIncomeExpensePipe } from '../../pipes/sort-income-expense.pipe';

@Component({
  selector: 'app-details',
  standalone: true,
  imports: [CommonModule, SortIncomeExpensePipe],
  templateUrl: './details.component.html',
  styles: [],
})
export class DetailsComponent implements OnInit {
  incomeExpenseItems$!: Signal<IncomeExpense[]>;
  isLoading!: Signal<boolean>;

  constructor(private store: Store) {}

  ngOnInit(): void {
    this.incomeExpenseItems$ = this.store.selectSignal(incomeExpenseFeature.selectItems);
    this.isLoading = this.store.selectSignal(incomeExpenseFeature.selectLoading);
  }

  delete(itemId: string | null | undefined): void {
    if (!itemId) return;
    const incomeExpense = this.incomeExpenseItems$();
    this.store.dispatch(DetailsActions.deleteItem({ incomeExpense, itemId }));
  }
}
