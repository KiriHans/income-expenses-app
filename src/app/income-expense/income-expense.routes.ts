import { Routes } from '@angular/router';
import { IncomeExpenseComponent } from './components/income-expense.component';
import { DetailsComponent } from './components/details/details.component';
import { StatisticsComponent } from './components/statistics/statistics.component';

export default [
  {
    path: 'income-expenses',
    component: IncomeExpenseComponent,
  },
  {
    path: 'detail',
    component: DetailsComponent,
  },
  {
    path: '',
    pathMatch: 'full',
    component: StatisticsComponent,
  },
] satisfies Routes;
