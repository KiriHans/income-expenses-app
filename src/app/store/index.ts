import { isDevMode } from '@angular/core';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as ui from '../shared/store/ui.reducer';
import * as auth from './reducers/auth.reducer';
import * as incomeExpense from './reducers/income-expense.reducer';

export const stateFeatureKey = 'state';

export interface AppState {
  [ui.uiFeatureKey]: ui.UIState;
  [auth.authFeatureKey]: auth.State;
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.State;
}

export const reducers: ActionReducerMap<AppState> = {
  [ui.uiFeatureKey]: ui.uiFeature.reducer,
  [auth.authFeatureKey]: auth.userReducer,
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.reducer,
};

export const metaReducers: MetaReducer<AppState>[] = isDevMode() ? [] : [];
