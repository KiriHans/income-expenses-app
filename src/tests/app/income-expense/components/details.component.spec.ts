import { ComponentFixture, TestBed } from '@angular/core/testing';

import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { DetailsComponent } from 'src/app/income-expense/components/details/details.component';
import { IncomeExpense } from 'src/app/core/models/income-expenses.model';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { DetailsActions } from 'src/app/store/actions/income-expense.actions';

describe('DetailsComponent', () => {
  let component: DetailsComponent;
  let fixture: ComponentFixture<DetailsComponent>;

  let el: DebugElement;

  let compiled: HTMLElement;

  let store: MockStore;

  const initialState = {
    incomeExpense: {
      items: [new IncomeExpense('test', 100, 'income', 'itemId')],
      loading: false,
    },
  } as const;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [DetailsComponent],
      providers: [provideMockStore({ initialState })],
    });
    fixture = TestBed.createComponent(DetailsComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;

    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should countain a title', () => {
    const title = el.query(By.css('h4.card-title'));
    expect(title.nativeElement.textContent).toBe('Income expenses details');
  });

  it('should set items and loading variables on component initialization', () => {
    fixture = TestBed.createComponent(DetailsComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;

    expect(component.incomeExpenseItems$).toBeUndefined();
    expect(component.isLoading).toBeUndefined();

    component.ngOnInit();

    expect(component.incomeExpenseItems$()).toBe(initialState.incomeExpense.items);
    expect(component.isLoading()).toBe(initialState.incomeExpense.loading);
  });

  it('should countain a table with', () => {
    const table = el.query(By.css('table.table'));
    const columns = table.queryAll(By.css('th'));
    const expectedColumns = ['Description', 'Amount', 'Type', 'Actions'];

    columns.forEach((column, i) => {
      expect(column.nativeElement.textContent).toBe(expectedColumns[i]);
    });
  });

  it('should render the items from the store in the table', () => {
    const newState = {
      incomeExpense: {
        items: [
          new IncomeExpense('Item 1', 100, 'income', '1'),
          new IncomeExpense('Item 2', 200, 'income', '2'),
          new IncomeExpense('Item 3', 400, 'expense', '3'),
        ],
        loading: false,
      },
    };
    store.setState(newState);
    store.refreshState();
    fixture.detectChanges();

    const tableElements = el.queryAll(By.css('table tbody tr'));

    expect(tableElements).toHaveLength(newState.incomeExpense.items.length);

    tableElements.forEach((element, i) => {
      const [description, amount, type, button] = element.queryAll(By.css('td'));
      const item = newState.incomeExpense.items[i];

      expect(description.nativeElement.textContent).toBe(item.description);
      expect(amount.nativeElement.textContent).toBe(`${item.amount}`);
      expect(type.nativeElement.textContent.trim()).toBe(item.type);
      expect(button.nativeElement.textContent.trim()).toBe('Delete');
    });
  });

  it("Table element should have it's respective class in their type column ", () => {
    let tableElementType = el.queryAll(By.css('table tbody td'))[2];
    const item = { ...initialState.incomeExpense.items[0] };

    expect(tableElementType.nativeElement.textContent.trim()).toBe('income');
    expect(tableElementType.classes['text-success']).toBeTruthy();

    item.type = 'expense';

    const newState = {
      incomeExpense: {
        items: [item],
        loading: false,
      },
    };
    store.setState(newState);
    store.refreshState();
    fixture.detectChanges();

    tableElementType = el.queryAll(By.css('table tbody td'))[2];

    expect(tableElementType.nativeElement.textContent.trim()).toBe('expense');
    expect(tableElementType.classes['text-danger']).toBeTruthy();
  });

  it('should trigger delete method when button is pressed', () => {
    jest.spyOn(component, 'delete');
    const button = el.query(By.css('table button.btn-danger'));
    button.nativeElement.click();

    expect(component.delete).toHaveBeenCalled();
  });

  it('delete method should dispatch its action', () => {
    const incomeExpense = component.incomeExpenseItems$();
    const itemId = incomeExpense[0].id || '';
    jest.spyOn(store, 'dispatch');
    component.delete(itemId);

    expect(store.dispatch).toHaveBeenCalledWith(
      DetailsActions.deleteItem({ incomeExpense, itemId })
    );
  });

  it('delete method should not dispatch any action if item id does not exist', () => {
    const item = { ...initialState.incomeExpense.items[0] };
    item.id = undefined;

    jest.spyOn(store, 'dispatch');

    const newState = {
      incomeExpense: {
        items: [item],
        loading: false,
      },
    };
    store.setState(newState);
    store.refreshState();
    fixture.detectChanges();

    const button = el.query(By.css('table button.btn-danger'));
    button.nativeElement.click();

    expect(store.dispatch).not.toHaveBeenCalled();
  });
});
