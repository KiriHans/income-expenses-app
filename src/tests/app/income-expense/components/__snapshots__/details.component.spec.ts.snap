// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h4
            class="card-title"
          >
            Income expenses details
          </h4>
          <table
            class="table"
          >
            <thead>
              <tr>
                <th>
                  Description
                </th>
                <th>
                  Amount
                </th>
                <th>
                  Type
                </th>
                <th>
                  Actions
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  test
                </td>
                <td>
                  100
                </td>
                <td
                  class="text-success"
                >
                   income 
                </td>
                <td>
                  <button
                    class="btn btn-danger"
                    type="button"
                  >
                     Delete 
                  </button>
                </td>
              </tr>
              
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
